snapcraft/internal/project_loader/_extensions/_utils.py

```diff
@@ -152,44 +152,91 @@ def _apply_extension(
             )
 
-    # Next, apply the part-specific components
+    # Next, apply the part-specific components.
     part_extension = extension.part_snippet
     parts = yaml_data["parts"]
     for part_name, part_definition in parts.items():
+
         for property_name, property_value in part_extension.items():
             part_definition[property_name] = _apply_extension_property(
                 part_definition.get(property_name), property_value
             )
 
-    # Finally, add any parts specified in the extension
+        # Stores the extension's list of part_snippets in each part.
+        parts[part_name] = part_definition
+
+    # Finally, add any parts specified in the extension.
     for part_name, part_definition in extension.parts.items():
         # If a extension part name clashes with a part that already exists, error.
         if part_name in parts:
             raise errors.ExtensionPartConflictError(extension_name, part_name)
 
+        # Stores the extension's list of parts in the parts section.
         parts[part_name] = part_definition
 
 
-def _apply_extension_property(existing_property: Any, extension_property: Any):
-    if existing_property:
-        # If the property is not scalar, merge them
+def _apply_extension_property(existing_property: Any, extension_property: Any) -> Dict[str, Any]:
+    """Take the user-defined yaml properties and apply any missing
+    extension properties. If there is a property defined in both, 
+    the user-defined property is applied.
+    """
+
+    # If there is no user-defined property, then just add the extension-defined property. 
+    if not existing_property:
+        return extension_property
+
+    # If there is a user-defined property, then take care when merging with the extension-defined property.
+    elif existing_property:
+
         if isinstance(existing_property, list) and isinstance(extension_property, list):
-            return _merge_lists(existing_property, extension_property)
+
+            # Additional check if the existing_property is a list of OrderedDicts.
+            # If there is a build-environment section defined in a user's part, it
+            # is a list of OrderedDicts that needs to be merged with the gnome-3-34
+            # extension build-environment section which is a list of dicts. The list
+            # of order OrderedDicts needs to be made into a list of dicts so we can
+            # keep track of each build-environment var to make sure we use the user-
+            # defined variable. 
+            # See https://bugs.launchpad.net/snapcraft/+bug/1853040 for more info.
+            temp_list = []
+            seen = set()
+
+            # Safely copy the list of OrderedDicts to a list of dicts.
+            for item in existing_property:
+                if isinstance(item, collections.OrderedDict):
+                    temp_list.append(dict(item))
+                if len(temp_list) > 0:
```

Comment:
>Is this indent correct?

Yes, this indent is correct. This for loop is supposed to only be triggered if we're dealing with an existing_property that is a list and an extension_property that is a list.

>I think you would lose data if existing_property was a list containing non-ordereddict.

I was just being cautious here... maybe one day we have to deal with a list with some combination of dicts and ordereddicts. I wanted all ordereddicts to be put into a dict, one item at a time. Alternatively, we could make the assumption that if there is a list whose first item is an ordereddict, then all of the lists items should be ordereddicts and replace this block with:

```
if isinstance(existing_property[0], collections.OrderedDict):
    for item in existing_property:
        temp_list.append(dict(item))
    if len(temp_list) > 0:
        existing_property = temp_list
```

Now, that being said, this block isn't needed at all if we go the route of making the build-environment an ordereddict like you have in [cjp256@576ccae](https://github.com/cjp256/snapcraft/commit/576ccae204c5784e01b2907bc8698aed12ee8504). I actually like your solution better because it reads better and now all lists are of ordereddicts (well, or strings) and we don't have a mixture of lists of ordered dicts and lists of dicts.